komorebi workspaces widget stops responding to clicks after display sleep/wake (stale cached monitor_hwnd)

### Windows version

Windows 11

### Windows OS build

26200.8875 (25H2)

### YASB version installed

YASB Reborn v2.0.5 x64 (stable), YASB-CLI v1.1.7

### Describe the bug

After the monitors sleep and wake, the `komorebi.workspaces` widget stops responding to clicks. The bar renders fine and the workspace buttons still highlight as the focused workspace changes, but clicking a button does nothing and each click logs `TypeError: 'NoneType' object is not subscriptable`.

komorebi itself is healthy throughout — `komorebic state` responds normally and its monitor list is correct. Only YASB needs restarting; `yasbc reload` fixes it until the next sleep.

**Steps to reproduce**

1. Run YASB with a `komorebi.workspaces` widget on a bar.
2. Let the displays sleep (or turn them off), then wake them.
3. Click any workspace button.

**Root cause**

`_komorebi_screen` is `None` because the widget is resolving its screen with a monitor handle that no longer exists.

- `src/core/bar.py#L105` caches the bar's monitor handle once in `__init__` and hands it to every widget at `#L262`:

  ```python
  self.monitor_hwnd = get_monitor_hwnd(int(self.winId()))   # L105, __init__ only
  ...
  widget.monitor_hwnd = self.monitor_hwnd                   # L262, _add_widgets only
  ```

- Windows reissues `HMONITOR` handles when the displays come back, so that cached value goes stale. `on_geometry_changed` (`bar.py#L152`) calls `position_bar()` and `update_app_bar()` but never recomputes `monitor_hwnd` — and it wouldn't help here anyway, since the resolution is unchanged and `geometryChanged` doesn't necessarily fire on a wake.

- `src/core/widgets/komorebi/workspaces.py#L403` then prefers that stale cached value over a live lookup:

  ```python
  self._screen_hwnd = self.monitor_hwnd or get_monitor_hwnd(int(QWidget.winId(self)))
  ```

  Because `self.monitor_hwnd` is always truthy, the live lookup is never reached, `get_screen_by_hwnd` (`#L406`) returns `None`, and the widget can never recover on its own. Every subsequent click hits `self.parent_widget._komorebi_screen["index"]` (`#L79` and `#L197`) and throws.

I confirmed the handle mismatch is the mechanism: while the bar was broken, the live `HMONITOR` values enumerated via `EnumDisplayMonitors` (`69491`, `131245`) matched the monitor `id`s in `komorebic state` exactly — so komorebi had re-enumerated correctly and YASB's private cached copy was the only stale one.

**Why the sibling widgets are unaffected**

`komorebi/active_layout.py#L326` and `komorebi/stack.py#L312` do the same job without the cache, and they keep working across a wake:

```python
self._screen_hwnd = get_monitor_hwnd(int(QWidget.winId(self)))
```

`workspaces.py` is the only komorebi widget using the `self.monitor_hwnd or ...` form.

**Regression**

The `self.monitor_hwnd or` prefix was introduced in 91c6fa3 ("fix(workspace): prevent komorebi widget freeze and broken states after auto-hide", #902), first shipped in **v2.0.1**. The intent makes sense — a hidden/auto-hidden widget's `winId()` yields no monitor, so falling back to the bar's handle keeps the widget working. The problem is only the precedence: the cached value wins unconditionally, including when the live lookup would have returned a correct, current handle.

Worth noting I don't use `auto_hide` at all, so on this config the cached path provides no benefit and only introduces the staleness.

**Suggested fix**

Inverting the precedence in `workspaces.py#L403` looks like it addresses this without regressing #902:

```python
self._screen_hwnd = get_monitor_hwnd(int(QWidget.winId(self))) or self.monitor_hwnd
```

`get_monitor_hwnd` (`src/core/utils/win32/utils.py#L90`) returns `None` cleanly when `MonitorFromWindow` finds nothing, which is exactly the hidden-widget case #902 was fixing — so the cached bar handle still covers it, while a visible widget always resolves against the current handle and self-heals after a wake.

Happy to open a PR for this if you'd like it, and to test any alternative you'd prefer (e.g. refreshing `bar.monitor_hwnd` and re-propagating to widgets instead).

I'd also flag that the same cached `monitor_hwnd` is consumed by `glazewm/workspaces.py`, `yasb/active_window.py`, `brightness.py`, `control_center.py`, `grouper.py`, `taskbar.py` and `window_controls.py` — I haven't tested those across a wake, but any of them that resolves a monitor from it may have the same staleness.

### Relevant log output

```shell
2026-08-06 12:02:55,186 [ERROR] [MainThread] [root/workspaces.py:199]: Failed to focus workspace at index 0
Traceback (most recent call last):
  File "D:\a\yasb\yasb\src\core\widgets\komorebi\workspaces.py", line 197, in activate_workspace
TypeError: 'NoneType' object is not subscriptable
2026-08-06 12:02:57,536 [ERROR] [MainThread] [root/workspaces.py:199]: Failed to focus workspace at index 0
Traceback (most recent call last):
  File "D:\a\yasb\yasb\src\core\widgets\komorebi\workspaces.py", line 197, in activate_workspace
TypeError: 'NoneType' object is not subscriptable
2026-08-06 12:04:34,696 [ERROR] [MainThread] [root/workspaces.py:199]: Failed to focus workspace at index 4
Traceback (most recent call last):
  File "D:\a\yasb\yasb\src\core\widgets\komorebi\workspaces.py", line 197, in activate_workspace
TypeError: 'NoneType' object is not subscriptable

# Windows event log for the same window, showing the wake that preceded the first error:
#   11:53:05  Kernel-Power 566  session 0 -> 1  (displays sleep, NextSessionType=1)
#   12:02:24  Kernel-Power 566  session 1 -> 3  (displays wake,  NextSessionType=0)
#   12:02:55  first "Failed to focus workspace"

# After `yasbc reload` the widget works again until the next sleep:
2026-08-06 12:08:13,763 [INFO] [MainThread] [root/bar_manager.py:103]: Starting KomorebiEventListener...
2026-08-06 12:08:13,905 [INFO] [Dummy-9] [root/event_listener.py:147]: Komorebi connected to named pipe: yasb-68c0d2de-91b9-11f1-8af7-7c10c967c7cf
```

--- a/src/core/widgets/komorebi/workspaces.py
+++ b/src/core/widgets/komorebi/workspaces.py
@@ -76,7 +76,10 @@ class WorkspaceButton(QPushButton):
 
     def activate_workspace(self):
         try:
-            self.komorebic.activate_workspace(self.parent_widget._komorebi_screen["index"], self.workspace_index)
+            screen = self.parent_widget._komorebi_screen
+            if screen is None:
+                return
+            self.komorebic.activate_workspace(screen["index"], self.workspace_index)
         except Exception:
             logging.exception("Failed to focus workspace at index %s", self.workspace_index)
 
@@ -194,7 +197,10 @@ class WorkspaceButtonWithIcons(QFrame):
 
     def activate_workspace(self):
         try:
-            self.komorebic.activate_workspace(self.parent_widget._komorebi_screen["index"], self.workspace_index)
+            screen = self.parent_widget._komorebi_screen
+            if screen is None:
+                return
+            self.komorebic.activate_workspace(screen["index"], self.workspace_index)
         except Exception:
             logging.exception("Failed to focus workspace at index %s", self.workspace_index)
 
@@ -400,7 +406,9 @@ class WorkspaceWidget(BaseWidget):
 
     def _update_komorebi_state(self, komorebi_state: dict) -> bool:
         try:
-            self._screen_hwnd = self.monitor_hwnd or get_monitor_hwnd(int(QWidget.winId(self)))
+            self._screen_hwnd = get_monitor_hwnd(int(QWidget.winId(self)))
+            if self._screen_hwnd is None:
+                return False
             self._komorebi_state = komorebi_state
             if self._komorebi_state:
                 self._komorebi_screen = self._komorebic.get_screen_by_hwnd(self._komorebi_state, self._screen_hwnd)

--- a/src/core/widgets/yasb/active_window.py
+++ b/src/core/widgets/yasb/active_window.py
@@ -8,12 +8,12 @@ import win32process
 from PIL import Image
 from PyQt6.QtCore import QElapsedTimer, Qt, QTimer, pyqtSignal
 from PyQt6.QtGui import QImage, QPixmap
-from PyQt6.QtWidgets import QLabel
+from PyQt6.QtWidgets import QLabel, QWidget
 
 from core.events.service import EventService
 from core.events.win32 import WinEvent
 from core.utils.win32.app_icons import get_window_icon
-from core.utils.win32.utils import get_app_name_from_aumid, get_app_name_from_pid, get_hwnd_info
+from core.utils.win32.utils import get_app_name_from_aumid, get_app_name_from_pid, get_hwnd_info, get_monitor_hwnd
 from core.validation.widgets.yasb.active_window import ActiveWindowConfig
 from core.widgets.base import BaseWidget
 from settings import APP_BAR_TITLE
@@ -226,11 +226,12 @@ class ActiveWindowWidget(BaseWidget):
             return
 
         monitor_name = win_info["monitor_info"].get("device", None)
+        widget_monitor = get_monitor_hwnd(int(QWidget.winId(self)))
 
         if (
             self.config.monitor_exclusive
             and self.screen().name() != monitor_name
-            and win_info.get("monitor_hwnd", "Unknown") != self.monitor_hwnd
+            and win_info.get("monitor_hwnd", "Unknown") != widget_monitor
         ):
             self._set_no_window_or_hide()
         else:
@@ -284,13 +285,6 @@ class ActiveWindowWidget(BaseWidget):
             if is_uwp:
                 try:
                     parent_pid = process["pid"]
-
-                    def _find_real_pid(child_hwnd, _):
-                        _, child_pid = win32process.GetWindowThreadProcessId(child_hwnd)
-                        if child_pid and child_pid != parent_pid:
-                            return False
-                        return True
-
                     found_pids = []
 
                     def _collect_pids(child_hwnd, _):

--- a/src/core/widgets/yasb/brightness.py
+++ b/src/core/widgets/yasb/brightness.py
@@ -3,11 +3,12 @@ from datetime import datetime
 
 from PyQt6.QtCore import QEvent, QRect, Qt, QTimer
 from PyQt6.QtGui import QShowEvent, QWheelEvent
-from PyQt6.QtWidgets import QFrame, QHBoxLayout, QLabel, QSlider, QStyle, QStyleOptionSlider, QVBoxLayout
+from PyQt6.QtWidgets import QFrame, QHBoxLayout, QLabel, QSlider, QStyle, QStyleOptionSlider, QVBoxLayout, QWidget
 
 from core.utils.qobject import is_valid_qobject
 from core.utils.tooltip import CustomToolTip, set_tooltip
 from core.utils.utilities import PopupWidget, build_progress_widget
+from core.utils.win32.utils import get_monitor_hwnd
 from core.validation.widgets.yasb.brightness import BrightnessConfig
 from core.widgets.base import BaseWidget
 from core.widgets.services.brightness.service import BrightnessService
@@ -57,8 +58,7 @@ class BrightnessWidget(BaseWidget):
 
     @property
     def _hmonitor(self) -> int | None:
-        """Bar-resolved monitor handle (set by Bar after position)."""
-        return self.monitor_hwnd
+        return get_monitor_hwnd(int(QWidget.winId(self)))
 
     def showEvent(self, a0: QShowEvent | None):
         """Handle widget show event detect monitor and check support."""

--- a/src/core/widgets/yasb/control_center.py
+++ b/src/core/widgets/yasb/control_center.py
@@ -3,6 +3,7 @@ from PyQt6.QtWidgets import QVBoxLayout, QWidget
 
 from core.utils.qobject import is_valid_qobject
 from core.utils.utilities import PopupWidget
+from core.utils.win32.utils import get_monitor_hwnd
 from core.validation.widgets.yasb.control_center import ControlCenterConfig
 from core.widgets.base import BaseWidget
 from core.widgets.services.brightness.service import BrightnessService
@@ -65,7 +66,7 @@ class ControlCenterWidget(BaseWidget):
 
     @property
     def _hmonitor(self) -> int | None:
-        return self.monitor_hwnd
+        return get_monitor_hwnd(int(QWidget.winId(self)))
 
     def _audio_services(self) -> dict[str, object]:
         return {

--- a/src/core/widgets/yasb/grouper.py
+++ b/src/core/widgets/yasb/grouper.py
@@ -2,9 +2,10 @@ import logging
 from typing import Any
 
 from PyQt6.QtCore import QEvent
-from PyQt6.QtWidgets import QFrame, QHBoxLayout, QPushButton
+from PyQt6.QtWidgets import QFrame, QHBoxLayout, QPushButton, QWidget
 
 from core.utils.widget_builder import WidgetBuilder
+from core.utils.win32.utils import get_monitor_hwnd
 from core.validation.widgets.yasb.grouper import GrouperWidgetConfig
 from core.widgets.base import BaseWidget
 
@@ -130,10 +131,13 @@ class GrouperWidget(BaseWidget):
     def _propagate_bar_context(self) -> None:
         """Propagate bar context to existing child widgets."""
         try:
+            monitor_hwnd = get_monitor_hwnd(int(QWidget.winId(self)))
+            if monitor_hwnd is None:
+                return
             for cw in self._child_widgets:
                 try:
                     cw.bar_id = self.bar_id
-                    cw.monitor_hwnd = self.monitor_hwnd
+                    cw.monitor_hwnd = monitor_hwnd
                     cw.screen_name = self.screen_name
                     cw.parent_layout_type = getattr(self, "parent_layout_type", None)
                 except Exception:

--- a/src/core/widgets/yasb/window_controls.py
+++ b/src/core/widgets/yasb/window_controls.py
@@ -5,11 +5,11 @@ import win32api
 import win32gui
 import win32process
 from PyQt6.QtCore import QEasingCurve, QPropertyAnimation, QTimer
-from PyQt6.QtWidgets import QGraphicsOpacityEffect, QLabel, QPushButton
+from PyQt6.QtWidgets import QGraphicsOpacityEffect, QLabel, QPushButton, QWidget
 
 from core.utils.tooltip import set_tooltip
 from core.utils.utilities import refresh_widget_style
-from core.utils.win32.utils import get_app_name_from_pid, is_window_maximized
+from core.utils.win32.utils import get_app_name_from_pid, get_monitor_hwnd, is_window_maximized
 from core.utils.win32.window_actions import (
     close_application,
     maximize_window,
@@ -43,13 +43,13 @@ except ImportError:
 class _ForegroundPollResult:
     """Lightweight container for a single poll cycle result."""
 
-    __slots__ = ("hwnd", "valid", "maximized", "monitor_hwnd", "app_name")
+    __slots__ = ("hwnd", "valid", "maximized", "fg_monitor_hwnd", "app_name")
 
-    def __init__(self, hwnd: int, valid: bool, maximized: bool, monitor_hwnd: int, app_name: str):
+    def __init__(self, hwnd: int, valid: bool, maximized: bool, fg_monitor_hwnd: int, app_name: str):
         self.hwnd = hwnd
         self.valid = valid
         self.maximized = maximized
-        self.monitor_hwnd = monitor_hwnd
+        self.fg_monitor_hwnd = fg_monitor_hwnd
         self.app_name = app_name
 
 
@@ -115,15 +115,15 @@ class _ForegroundPoller:
             if hwnd == self._last_hwnd and self._last_result is not None and self._last_result.valid:
                 try:
                     maximized = is_window_maximized(hwnd)
-                    monitor_hwnd = int(win32api.MonitorFromWindow(hwnd))
+                    fg_monitor_hwnd = int(win32api.MonitorFromWindow(hwnd))
                 except Exception:
                     # Window was likely destroyed
                     self._last_hwnd = 0
                     self._broadcast(_ForegroundPollResult(0, False, False, 0, ""))
                     return
-                if maximized != self._last_result.maximized or monitor_hwnd != self._last_result.monitor_hwnd:
+                if maximized != self._last_result.maximized or fg_monitor_hwnd != self._last_result.fg_monitor_hwnd:
                     self._broadcast(
-                        _ForegroundPollResult(hwnd, True, maximized, monitor_hwnd, self._last_result.app_name)
+                        _ForegroundPollResult(hwnd, True, maximized, fg_monitor_hwnd, self._last_result.app_name)
                     )
                 return
 
@@ -140,7 +140,7 @@ class _ForegroundPoller:
                 return
 
             maximized = is_window_maximized(hwnd)
-            monitor_hwnd = int(win32api.MonitorFromWindow(hwnd))
+            fg_monitor_hwnd = int(win32api.MonitorFromWindow(hwnd))
 
             # For UWP apps, ApplicationFrameHost.exe owns the window find the real child PID
             pid_for_name = pid
@@ -167,7 +167,7 @@ class _ForegroundPoller:
                 else:
                     self._app_name_cache[pid_for_name] = app_name
 
-            result = _ForegroundPollResult(hwnd, True, maximized, monitor_hwnd, app_name)
+            result = _ForegroundPollResult(hwnd, True, maximized, fg_monitor_hwnd, app_name)
             self._broadcast(result)
         except Exception:
             pass
@@ -293,8 +293,9 @@ class WindowControlsWidget(BaseWidget):
         # Per-widget monitor exclusivity check
         if self.config.monitor_exclusive:
             try:
-                if result.monitor_hwnd != self.monitor_hwnd:
-                    mon_info = win32api.GetMonitorInfo(result.monitor_hwnd)
+                widget_monitor = get_monitor_hwnd(int(QWidget.winId(self)))
+                if result.fg_monitor_hwnd != widget_monitor:
+                    mon_info = win32api.GetMonitorInfo(result.fg_monitor_hwnd)
                     if self.screen().name() != mon_info.get("Device"):
                         if self._is_visible:
                             self._tracked_hwnd = None

